### PR TITLE
Preprocess custom functions

### PR DESCRIPTION
**Pull request by pjferrell**
Add support for defining custom preprocessing function references before or after the normal generated preprocessing functions.  

Example was updated to reflect the new options.

After this PR is merged, the atlas-gentool will need to be updated with a new version of protoc-gen-preprocess.

### CODE REVIEW

**Comment by shishir1407** [comment]
@askurydzin Can you please review this PR?

File: options/preprocess.proto [hunk 1]
```diff
@@ -7,16 +7,18 @@ import "google/protobuf/descriptor.proto";
...
 // Message level options
 extend google.protobuf.MessageOptions{
-    PreprocessMessageOptions each = 11111;
+    PreprocessMessageOptions message = 11110;
+    PreprocessFieldOptions each = 11111;
 }
 
 message PreprocessMessageOptions {
-    PreprocessString string = 1;
+    string prefunction = 1;
```
Comment (askurydzin) [inline options/preprocess.proto:15]: I think breaking existing interface is not good. PreprocessString is a set of preprocessing actions that is applied to any string in the message.

So should you create:

```
message PreprocessFunction {
    string pre = 1;
    string post = 2;
}

...

PreprocessFunction function (func?) = 2;
```

File: main.go [hunk 2]
```diff
@@ -134,13 +146,17 @@ func generatePreprocessCall(g *protogen.GeneratedFile, varName string, repeated
 
 	if repeated {
 		g.P(`for _, v := range `, varName, `{`)
-		g.P(`if v != nil {`)
-		g.P(`v.Preprocess()`)
-		g.P(`}`)
+		g.P(`  if v != nil {`)
```
Comment (askurydzin) [inline main.go:149]: No need to add extra space, it's all to be run through the go formatter. here and everywere.

File: options/preprocess.proto (hunk 1 above)
Comment (askurydzin) [reply options/preprocess.proto:15]: Naming seems to be awkward anyway, why not post_function, so better to pack this into the separate message.

**Review by mphillips-infoblox — APPROVED** [review]

File: options/preprocess.proto (hunk 1 above)
Comment (pjferrell) [reply options/preprocess.proto]: Discussed and addressed in commit.

**Pull request merged**